crates/symbolicator/src/services/symbolication/process_minidump.rs

```diff
@@ -130,81 +128,187 @@ impl MinidumpState {
+    cfi_status: ObjectFileStatus,
+    /// Call frame information for the module, loaded as a [`SymbolFile`].
+    symbol_file: Option<SymbolFile>,
+    /// The DIF object candidates for for this module.
+    cfi_candidates: AllObjectCandidates,
 }
 
+/// A [`SymbolProvider`] that uses a [`CfiCacheActor`] to fetch
+/// CFI for stackwalking.
+///
+/// An instance of this type is always used to stackwalk one particular minidump.
 struct SymbolicatorSymbolProvider {
-    files: BTreeMap<DebugId, SymbolFile>,
-    missing_ids: RwLock<BTreeSet<DebugId>>,
+    /// The scope of the stackwalking request.
+    scope: Scope,
+    /// The sources from which to fetch CFI.
+    sources: Arc<[SourceConfig]>,
+    /// The object type of the minidump to stackwalk.
+    object_type: ObjectType,
+    /// The actor used for fetching CFI.
+    cficache_actor: CfiCacheActor,
+    /// An internal database of loaded CFI.
+    ///
+    /// The key consists of a module's debug identifier and base address.
+    cficaches: RwLock<HashMap<(DebugId, u64), CfiModule>>,
 }
 
 impl SymbolicatorSymbolProvider {
-    /// Load the CFI information from the cache.
-    ///
-    /// This reads the CFI caches from disk and returns them in a format suitable for the
-    /// processor to stackwalk.
-    pub fn new<'a, M: Iterator<Item = &'a (DebugId, PathBuf)>>(modules: M) -> Self {
+    pub fn new(
+        scope: Scope,
+        sources: Arc<[SourceConfig]>,
+        cficache_actor: CfiCacheActor,
+        object_type: ObjectType,
+    ) -> Self {
         Self {
-            files: modules
-                .filter_map(|(id, path)| Some((*id, Self::load(path)?)))
-                .collect(),
-
-            missing_ids: RwLock::new(BTreeSet::new()),
+            scope,
+            sources,
+            cficache_actor,
+            cficaches: Default::default(),
+            object_type,
         }
     }
 
-    fn load(cfi_path: &Path) -> Option<SymbolFile> {
-        sentry::with_scope(
-            |scope| {
-                scope.set_extra(
-                    "cfi_cache",
-                    sentry::protocol::Value::String(cfi_path.to_string_lossy().to_string()),
-                )
-            },
-            || {
-                let bytes = ByteView::open(cfi_path)
-                    .map_err(|err| {
-                        let stderr: &dyn std::error::Error = &err;
-                        tracing::error!(stderr, "Error while reading cficache");
-                    })
-                    .ok()?;
-
-                let cfi_cache = CfiCache::from_bytes(bytes)
-                    // This mostly never happens since we already checked the files
-                    // after downloading and they would have been tagged with
-                    // CacheStatus::Malformed.
-                    .map_err(|err| {
+    /// Fetches CFI for the given module, parses it into a `SymbolFile`, and stores it internally.
+    #[tracing::instrument(skip_all)]
```

Comment:
looking at some of our performance traces, I think this here is a bit overkill, as it is being called on *every* call to `walk_frame`, and in cases we already loaded the file, it will just early-return. Similarly, the `CfiCacheActor` may have stuff cached in the majority of cases anyways. The cache actor is smart enough to only start a new tracing span if it indeed fetches files from their source, which I think is what we are most interested in. So I would suggest to just remove this `instrument` annotation completely.

Reply:
Do you think it would make sense to open a span in the body of the function for the part that actually does something? Also, would it make sense to instrument `load_symbol_file`?